File: core/src/com/qtech/bubbleblaster/init/GameEvents.java
package com.qtech.bubbleblaster.init;

import com.qtech.bubbleblaster.QInternalAddon;
import com.qtech.bubbleblaster.common.gamestate.GameEvent;
import com.qtech.bubbleblaster.common.init.ObjectInit;
import com.qtech.bubbleblaster.registry.DeferredRegister;
import com.qtech.bubbleblaster.registry.Registers;
import com.qtech.bubbleblaster.registry.object.RegistryObject;
import com.qtech.bubbleblaster.state.BloodMoonEvent;

import java.util.function.Supplier;

/**
 * @see GameEvent
 */
@SuppressWarnings("unused")
public class GameEvents implements ObjectInit<GameEvent> {
    public static final DeferredRegister<GameEvent> GAME_EVENTS = DeferredRegister.create(QInternalAddon.ADDON_ID, Registers.GAME_EVENTS);

    // Bubbles
    public static final RegistryObject<BloodMoonEvent> BLOOD_MOON_EVENT = register("blood_moon", BloodMoonEvent::new);

    private static <T extends GameEvent> RegistryObject<T> register(String name, Supplier<T> supplier) {
        return GAME_EVENTS.register(name, supplier);
    }
}

File: src/main/kotlin/qtech/bubbles/init/GameEvents.kt
package qtech.bubbles.init

import qtech.bubbles.BBInternalMod
import qtech.bubbles.common.gamestate.GameEvent
import qtech.bubbles.common.init.ObjectInit
import qtech.bubbles.registry.DeferredRegister
import qtech.bubbles.registry.Registers
import qtech.bubbles.registry.`object`.RegistryObject
import qtech.bubbles.state.BloodMoonEvent
import java.util.function.Supplier

/**
 * @see GameEvent
 */
object GameEvents : ObjectInit<GameEvent> {
    @JvmField
    val GAME_EVENTS: DeferredRegister<GameEvent> = DeferredRegister.create(BBInternalMod.ADDON_ID, Registers.GAME_EVENTS)

    // Bubbles
    val BLOOD_MOON_EVENT: RegistryObject<BloodMoonEvent> = register("blood_moon") { BloodMoonEvent() }

    @Suppress("SameParameterValue")
    private fun <T : GameEvent> register(name: String, supplier: Supplier<T>): RegistryObject<T> {
        return GAME_EVENTS.register(name, supplier)
    }
}